In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import numpy as np

In [ ]:
!pwd

/content


In [ ]:
# about rating
path_rating = 'drive/MyDrive/Notebooks/ml-100k/u.data'
# about movie
path_movie = 'drive/MyDrive/Notebooks/ml-100k/u.item'
# about user
path_user = 'drive/MyDrive/Notebooks/ml-100k/u.user'

In [ ]:
def parse(s):
  seq = s.split(' | ')
  return list(map(lambda s: s.replace(' ', '_'), seq))

parse('user id | age | gender | occupation | zip code')

['user_id', 'age', 'gender', 'occupation', 'zip_code']

In [ ]:
user_col = parse('user id | age | gender | occupation | zip code')
print(user_col)
users = pd.read_csv(path_user, sep='|', names=user_col)
users = users.set_index(user_col[0])
print(users.shape)
display(users.head(5))

['user_id', 'age', 'gender', 'occupation', 'zip_code']
(943, 4)


age gender  occupation zip_code
user_id                                 
1         24      M  technician    85711
2         53      F       other    94043
3         23      M      writer    32067
4         24      M  technician    43537
5         33      F       other    15213

In [ ]:
# movie id | movie title | release date | video release date | IMDb URL | unknown | Action | Adventure | Animation | Children's | Comedy | Crime | Documentary | Drama | Fantasy | Film-Noir | Horror | Musical | Mystery | Romance | Sci-Fi | Thriller | War | Western

movie_col = parse("movie id | movie title | release date | video release date | IMDb URL | unknown | Action | Adventure | Animation | Children's | Comedy | Crime | Documentary | Drama | Fantasy | Film-Noir | Horror | Musical | Mystery | Romance | Sci-Fi | Thriller | War | Western")
print(movie_col)
movies = pd.read_csv(path_movie, sep='|', names=movie_col, encoding='ISO-8859-1')
movies = movies.set_index(movie_col[0])
print(movies.shape)
display(movies.head(5))

['movie_id', 'movie_title', 'release_date', 'video_release_date', 'IMDb_URL', 'unknown', 'Action', 'Adventure', 'Animation', "Children's", 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']
(1682, 23)


movie_title release_date  ...  War Western
movie_id                                  ...             
1          Toy Story (1995)  01-Jan-1995  ...    0       0
2          GoldenEye (1995)  01-Jan-1995  ...    0       0
3         Four Rooms (1995)  01-Jan-1995  ...    0       0
4         Get Shorty (1995)  01-Jan-1995  ...    0       0
5            Copycat (1995)  01-Jan-1995  ...    0       0

[5 rows x 23 columns]

In [ ]:
# user id | item id | rating | timestamp
rating_col = parse('user id | item id | rating | timestamp')
print(rating_col)
ratings = pd.read_csv(path_rating, sep='\t', names=rating_col, encoding='ISO-8859-1')
ratings = ratings.set_index(rating_col[0])
print(ratings.shape)
display(ratings.head(5))

['user_id', 'item_id', 'rating', 'timestamp']
(100000, 3)


item_id  rating  timestamp
user_id                            
196          242       3  881250949
186          302       3  891717742
22           377       1  878887116
244           51       2  880606923
166          346       1  886397596

**영화별로 그루핑하여 평점 출력**

In [ ]:
ratings_mean = ratings.groupby(rating_col[1]).mean()
print(ratings_mean.shape)
ratings_mean.sort_values(by=rating_col[2], ascending=False)

(1682, 2)


rating    timestamp
item_id                     
814         5.0  886302261.0
1599        5.0  880142614.0
1201        5.0  891383687.0
1122        5.0  883326498.0
1653        5.0  889489913.0
...         ...          ...
1568        1.0  885547222.0
1567        1.0  885547123.0
1566        1.0  885546248.0
1565        1.0  885549463.0
852         1.0  882396869.0

[1682 rows x 2 columns]

In [ ]:
movie_ratings_mean = ratings_mean.groupby(["item_id"])["rating"].mean()
print(movie_ratings_mean.shape)
display(movie_ratings_mean)

(1682,)


item_id
1       3.878319
2       3.206107
3       3.033333
4       3.550239
5       3.302326
          ...   
1678    1.000000
1679    3.000000
1680    2.000000
1681    3.000000
1682    3.000000
Name: rating, Length: 1682, dtype: float64

**평점을 내림차순으로 정렬하여 상위 5개 영화목록 출력**

In [ ]:
movie_rating_sort = movie_ratings_mean.sort_values(ascending=False)[:5]
movie_rating_sort

item_id
1293    5.0
1467    5.0
1653    5.0
814     5.0
1122    5.0
Name: rating, dtype: float64

In [ ]:
recom_movies = movies.loc[movie_rating_sort.index]
recom_movies['movie_title']

item_id
1293                                      Star Kid (1997)
1467                 Saint of Fort Washington, The (1993)
1653    Entertaining Angels: The Dorothy Day Story (1996)
814                         Great Day in Harlem, A (1994)
1122                       They Made Me a Criminal (1939)
Name: movie_title, dtype: object

오차 계산 : **RMSE (Root Mean Square Error)** 사용

In [ ]:
def RMSE(answer, prediction):
  e = (np.array(answer) - np.array(prediction)) ** 2
  return np.sqrt(np.mean((e)))

In [ ]:
rmse = []
for user in set(ratings.index):
  # 실제 해당 유저가 평가한 점수
  answer = ratings.loc[user]['rating']
  # 해당 유저가 평가한 영화의 평균 점수
  prediction = movie_ratings_mean.loc[ratings.loc[user]['item_id']]
  accuracy = RMSE(answer, prediction)
  rmse.append(accuracy)

print(np.mean(rmse))

0.996007224010567


**표본이 적을 경우 신뢰도가 떨어지므로** item_id 당 받은 평가의 갯수가 상위 1000위 이하인 영화들 중에서 추천 (전체 : 1682개)  


In [ ]:
ratings_1000 = ratings.groupby(['item_id']).size().sort_values(ascending=False)[:1000]
ratings_1000

item_id
50      583
258     509
100     508
181     507
294     485
       ... 
1419     16
854      16
453      16
997      16
256      16
Length: 1000, dtype: int64

In [ ]:
# 평가 갯수가 상위 1000위 안에 속하는 경우만 필터링
filtered_ratings = ratings[ratings['item_id'].isin(ratings_1000.index)]
filtered_ratings

item_id  rating  timestamp
user_id                            
196          242       3  881250949
186          302       3  891717742
244           51       2  880606923
166          346       1  886397596
298          474       4  884182806
...          ...     ...        ...
880          476       3  880175444
716          204       5  879795543
276         1090       1  874795795
13           225       2  882399156
12           203       3  879959583

[96056 rows x 3 columns]

In [ ]:
filtered_ratings_mean = filtered_ratings.groupby(['item_id'])['rating'].mean()
filtered_ratings_mean

item_id
1       3.878319
2       3.206107
3       3.033333
4       3.550239
5       3.302326
          ...   
1407    2.250000
1411    2.571429
1419    2.125000
1444    2.473684
1478    2.722222
Name: rating, Length: 1000, dtype: float64

In [ ]:
filtered_rating_mean_sort = filtered_ratings_mean.sort_values(ascending=False)[:10]
filtered_rating_mean_sort

item_id
408    4.491071
318    4.466443
169    4.466102
483    4.456790
114    4.447761
64     4.445230
603    4.387560
12     4.385768
50     4.358491
178    4.344000
Name: rating, dtype: float64

In [ ]:
movies.loc[filtered_rating_mean_sort.index]

movie_title  ... Western
item_id                                                     ...        
408                                  Close Shave, A (1995)  ...       0
318                                Schindler's List (1993)  ...       0
169                             Wrong Trousers, The (1993)  ...       0
483                                      Casablanca (1942)  ...       0
114      Wallace & Gromit: The Best of Aardman Animatio...  ...       0
64                        Shawshank Redemption, The (1994)  ...       0
603                                     Rear Window (1954)  ...       0
12                              Usual Suspects, The (1995)  ...       0
50                                        Star Wars (1977)  ...       0
178                                    12 Angry Men (1957)  ...       0

[10 rows x 23 columns]

현재까지는 사용자를 고려하지 않음. 단순히 표본이 많은 영화중 평점이 높은 순으로만 고려함.
